### ISSUE TITLE
Postgresql ARRAY not supported

### ISSUE THREAD

**Issue by rrottier** [fsprojects/SQLProvider#281]
### Description

I am trying to retrieve some data from a postgresql database using the SQL Provider.  I am able to query some tables without any issue but when I try to run a query on a table containing a column with a double precision [][] type.

The error is:
```
FsLab Experiment1.fsx(50,18): error FS3021: Unexpected exception from provided type 'FSharp.Data.Sql.SqlDataProvider,DatabaseVendor="2",ConnectionString="Server=crunch.****.com;Database=itrd_db;User Id=******;Password=*****",ResolutionPath="Z:\\FSharp\\EVTAnalysis\\EVTAnalysis\\packages\\Npgsql\\lib\\net45",UseOptionTypes="True"+dataContext+publicSchema+public.itrd_telemetrymatrix' member 'GetMethods': The type provider 'FSharp.Data.Sql.SqlTypeProvider' reported an error: Could not get columns for `public.itrd_telemetrymatrix`, the type `ARRAY` is unknown to Npgsql
```

But when I just run a plain Npgsql query I am able to get the data without any issue.  

### Related information 

* Windows 10
* Visual Studio 2015

**Comment by ingted** [comment]
Hi rrottier,

What version is your PG?
I bumped into some similar issues and after porting Npqsql 2.2.7 to .net 4.6.1 and replace Npgsql 3.1.4 in projects reference, problems were gone!

**Comment by rrottier** [comment]
Yep I am using Npgsql 3.1.4 which is just what nuget got me.  I am not sure how to port Npgsql to .net 4.6.1 as I am pretty new to .Net land.  So if you have any detailed steps that I could follow that would be great.

But I am curious, if the issue is with Npgsql why then am I able to get the data without any issue when just running it as an SQL query through Npgsql 3.1.4 directly?

**Comment by janno-p** [comment]
Hey!

The error message is actually misleading. It comes from SqlProvider, not Npgsql, so it shouldn't make statements about what Npgsql supports or not.

The real cause of the error is that currently SqlProvider doesn't handle complex Npgsql/PosgreSQL types like array, enum, composite or range which are supported by Npgsql.

**Comment by ingted** [comment]
Hi rrottier,

Please try this Npgsql 2.2.7 with .net 4.6.1
https://github.com/ingted/npgsql4.6.1


Regards
Anibal

**Comment by ingted** [comment]
Hi rrottier,

Excuse me, I found that SQLProvider doesn't support ARRAY...

Here is the code.... (in Providers.Postgresql.fs)

            //"array"                       , typemap<Array>                      ["Array"]
            //"composite"                   , typemap<obj>                        ["Composite"]
            //"enum"                        , typemap<obj>                        ["Enum"]
            //"range"                       , typemap<Array>                      ["Range"]

**Comment by rrottier** [comment]
Ah I see, the array code is all commented out.  Does that mean that it is broken for now but may be fixed in the future?

Any suggestions for what else to use for Postgres access that might support arrays and F# Query language?

I know that npgsql works but then I am back to doing all my queries as strings...

**Comment by xkris6x** [comment]
I have a problem with roundcube
Error: xframework does not support database provider postgres
Someone help me

**Comment by Thorium** [comment]
Closed as fixed.
